src/HypreLinearSystem.C

```diff
@@ -295,41 +826,416 @@ HypreLinearSystem::finalizeSolver()
   HYPRE_IJVectorGetObject(sln_, (void**)&(solver->parSln_));
 }
 
-void
-HypreLinearSystem::loadComplete()
+
+void HypreLinearSystem::fill_entity_to_row_mapping()
 {
-  // All algorithms have called sumInto and populated LHS/RHS. Now we are ready
-  // to finalize the matrix at the HYPRE end. However, before we do that we need
-  // to process unfilled rows and process them appropriately. Any row acted on
-  // by sumInto method will have toggled the rowFilled_ array to RS_FILLED
-  // status. Before finalizing assembly, we process rows that still have an
-  // RS_UNFILLED status and set their diagonal entries to 1.0 (dummy row)
-  //
-  // TODO: Alternate design to eliminate dummy rows. This will require
-  // load-balancing on HYPRE end.
-
-  HypreIntType hnrows = 1;
-  HypreIntType hncols = 1;
-  double getval;
-  double setval = 1.0;
-  for (HypreIntType i=0; i < numRows_; i++) {
-    if (rowFilled_[i] == RS_FILLED) continue;
-    HypreIntType lid = iLower_ + i;
-    HYPRE_IJMatrixGetValues(mat_, hnrows, &hncols, &lid, &lid, &getval);
-    if (std::fabs(getval) < 1.0e-12)
-      HYPRE_IJMatrixSetValues(mat_, hnrows, &hncols, &lid, &lid, &setval);
+  const stk::mesh::BulkData& bulk = realm_.bulk_data();
+  stk::mesh::Selector selector = bulk.mesh_meta_data().universal_part() & !(realm_.get_inactive_selector());
+  entityToLIDHost_ = HypreIntTypeViewHost("entityToRowLIDHost",bulk.get_size_of_entity_index_space());
+  entityToLID_ = HypreIntTypeView("entityToRowLID",bulk.get_size_of_entity_index_space());
+
+  const stk::mesh::BucketVector& nodeBuckets = realm_.get_buckets(stk::topology::NODE_RANK, selector);
+  for(const stk::mesh::Bucket* bptr : nodeBuckets) {
+    const stk::mesh::Bucket& b = *bptr;
+    for(size_t i=0; i<b.size(); ++i) {
+      stk::mesh::Entity node = b[i];
+      const auto naluId = *stk::mesh::field_data(*realm_.naluGlobalId_, node);
+      const auto mnode = bulk.get_entity(stk::topology::NODE_RANK, naluId);
```

Comment:
BulkData::get_entity is expensive (map find). I think you're getting the master node for periodic, meaning naluId==stkId. You could say something like 'mnode = (naluId == bulk.identifier(node)) ? node : bulk.get_entity...'. That way you only use get_entity when you really need to. BulkData::identifier is a lot faster than get_entity.

Reply:
This code is legacy from the original implementation though I'm still relying on it. Changing this line (and one other instance of the same call) to:
const auto mnode = (naluId == bulk.identifier(node)) ? node : bulk.get_entity(stk::topology::NODE_RANK, naluId);
seems to work though I can't tell if things are faster or not.